### PR TITLE
Allow subqueries in ANY expressions 

### PR DESCRIPTION
**Pull request by mxm**
This enables queries like the following:

`Select 2 where 1 = ANY(Select 1)` which were not possible beforehand.

### CODE REVIEW

File: sql/src/main/java/io/crate/executor/transport/TransportExecutor.java [hunk 1]
```diff
@@ -352,6 +352,7 @@ public Task visitMultiPhasePlan(MultiPhasePlan multiPhasePlan, final Void contex
                         consumer.accept(null, e);
                     }
                 });
+                SubSelectSymbolReplacer replacer = new SubSelectSymbolReplacer(rootPlan, entry.getValue());
```
Comment (mikethebeer) [inline sql/src/main/java/io/crate/executor/transport/TransportExecutor.java]: @mxm why is this movement necessary?

File: sql/src/main/java/io/crate/executor/transport/TransportExecutor.java [hunk 2]
```diff
@@ -91,6 +91,7 @@
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
```
Comment (mikethebeer) [inline sql/src/main/java/io/crate/executor/transport/TransportExecutor.java]: import necessary?

File: sql-parser/src/main/java/io/crate/sql/tree/SubqueryExpression.java [hunk 3]
```diff
@@ -25,6 +25,8 @@
     extends Expression {
     private final Query query;
 
+    private boolean isArrayExpression;
```
Comment (mikethebeer) [inline sql-parser/src/main/java/io/crate/sql/tree/SubqueryExpression.java]: AFAIK this is not an `ArrayExpression` it is a `SubqueryExpression`

**Comment by mikethebeer** [comment]
tests missing

File: sql/src/main/java/io/crate/executor/transport/TransportExecutor.java (hunk 1 above)
Comment (mxm) [inline sql/src/main/java/io/crate/executor/transport/TransportExecutor.java]: Code clarity. The `SubSelectSymbolReplacer` is only used in the next line and has no dependencies on any symbols above.

File: sql-parser/src/main/java/io/crate/sql/tree/SubqueryExpression.java (hunk 3 above)
Comment (mxm) [inline sql-parser/src/main/java/io/crate/sql/tree/SubqueryExpression.java]: This field denotes that the expression is part of an array expression.

Comment (mikethebeer) [inline sql-parser/src/main/java/io/crate/sql/tree/SubqueryExpression.java]: @mxm ok but currently the naming (as it is now) indicates that the argument is an `ArrayExpression` if the right-hand side is an instance of `SubqueryExpression`. This is IMHO wrong since there is a important distinction between an array/row comparison and subquery expressions.

Comment (mxm) [inline sql-parser/src/main/java/io/crate/sql/tree/SubqueryExpression.java]: The `SubqueryExpression` has to result in an array expression in an `ANY` operation. Otherwise it is not valid.

The following are interchangeable:

`select id where id = ANY([1])` <-- array argument
`select id where id = ANY(Select 1)` <-- array argument

File: sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java [hunk 4]
```diff
@@ -118,8 +118,8 @@ public RelationAnalyzer(ClusterService clusterService, Functions functions, Sche
 
     public AnalyzedRelation analyze(Node node, StatementAnalysisContext statementContext) {
         AnalyzedRelation relation = process(node, statementContext);
-        relation = SubselectRewriter.rewrite(relation);
-        return relationNormalizer.normalize(relation, statementContext.transactionContext());
+        AnalyzedRelation rewrittenRelation = SubselectRewriter.rewrite(relation);
```
Comment (matriv) [inline sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java]: I wouldn't keep this change, but your call.

File: sql/src/main/java/io/crate/operation/collect/sources/TableFunctionCollectSource.java [hunk 5]
```diff
@@ -34,7 +34,12 @@
 import io.crate.metadata.table.TableInfo;
 import io.crate.metadata.tablefunctions.TableFunctionImplementation;
 import io.crate.operation.InputFactory;
-import io.crate.operation.collect.*;
+import io.crate.operation.collect.CrateCollector;
```
Comment (matriv) [inline sql/src/main/java/io/crate/operation/collect/sources/TableFunctionCollectSource.java]: Plz revert these import changes since you didn't touch anything else in the file.

File: sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java (hunk 4 above)
Comment (mxm) [inline sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java]: IMHO it is helpful for debugging.

File: sql/src/main/java/io/crate/operation/collect/sources/TableFunctionCollectSource.java (hunk 5 above)
Comment (mxm) [inline sql/src/main/java/io/crate/operation/collect/sources/TableFunctionCollectSource.java]: I have a checkstyle PR in my sleeve which will take care of that :)

File: sql/src/test/java/io/crate/operation/operator/any/AnyEqOperatorTest.java [hunk 6]
```diff
@@ -104,4 +104,10 @@ public void testNormalizeSymbol() throws Exception {
         assertNormalize("42 = ANY([42])", isLiteral(true));
         assertNormalize("42 = ANY([41, 43, -42])", isLiteral(false));
     }
+
+    @Test
+    public void testExceptionForwarding() throws Exception {
```
Comment (matriv) [inline sql/src/test/java/io/crate/operation/operator/any/AnyEqOperatorTest.java]: rename to something like "testIncompatibleTypes"

File: CHANGES.txt [hunk 7]
```diff
@@ -27,6 +27,10 @@ Breaking Changes
 Changes
 =======
 
+ - Introduce support for single row subselects in ANY, e.g.::
+
+   Select id from t where id = ANY(Select id from t2)
```
Comment (matriv) [inline CHANGES.txt]: plz use capitals for the keywords: SELECT FROM WHERE and indent by 2 spaces from the parent line

File: CHANGES.txt [hunk 8]
```diff
@@ -33,6 +33,10 @@ Changes
 
  - Changed the Enterprise License missing check message, as well as lowering
    the severity from HIGH to LOW.
+   
+  - Introduce support for single row subselects in ANY, e.g.::
+
+      Select id from t where id = ANY(Select id from t2)
```
Comment (mikethebeer) [inline CHANGES.txt]: this is not a single row subquery -> do you mean single column subquery?

Comment (mikethebeer) [inline CHANGES.txt]: probably say `select 1 where 1 = any (select 1)`

Comment (mxm) [inline CHANGES.txt]: Assuming that `id` is a primary key, it returns a single row and a single column :)

Comment (mxm) [inline CHANGES.txt]: I wanted to provide a more real-world example, than `select 1 where 1 = any (select 1)`.

Comment (dobe) [inline CHANGES.txt]: this is no single row subselect it is a single column subselect

File: core/src/main/java/io/crate/types/TableType.java [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A type which represents a table retrieved by a subquery.
```
Comment (dobe) [inline core/src/main/java/io/crate/types/TableType.java]: this is not subquery related. it also is not IN or ANY related. it should be described in a generic way.

File: core/src/main/java/io/crate/types/TableType.java [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+ * May be used in an IN or ANY operation, e.g. select 1 where 1 = ANY(Select 1)
+ */
+public class TableType extends DataType<Collection> implements CollectionType, Streamer<Object> {
+
+    public static final int ID = 102;
+    private DataType<?> valueType;
```
Comment (dobe) [inline core/src/main/java/io/crate/types/TableType.java]: consider calling this rowType, but then...

actually this type is not a generic TableType it is a SingleColumn table type. it does not work with multiple columns. also it is not future prove since let's say we want to use a special complex type as type of the row in the future, the semantic is different because the valueType here is actually the valueType of the first column of each row

File: core/src/main/java/io/crate/types/TableType.java [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+        }
+    }
+
+    @Override
+    public void writeValueTo(StreamOutput out, Object value) throws IOException {
+        if (value == null) {
+            out.writeVInt(0);
+            return;
+        }
+        if (valueType.id() == SetType.ID || valueType.id() == ArrayType.ID) {
```
Comment (dobe) [inline core/src/main/java/io/crate/types/TableType.java]: why do we unfold collections in the serialization?

File: CHANGES.txt (hunk 8 above)
Comment (mxm) [inline CHANGES.txt]: It is a single row-single column subselect. Will rephrase it to `single-column`.

File: core/src/main/java/io/crate/types/TableType.java (hunk 9 above)
Comment (mxm) [inline core/src/main/java/io/crate/types/TableType.java]: You're right. The original plan was to make this the type for all subqueries but we decided to postpone this for now.

File: core/src/main/java/io/crate/types/TableType.java (hunk 11 above)
Comment (mxm) [inline core/src/main/java/io/crate/types/TableType.java]: This is to support the new `TableType` with the existing `SetType` or `ArrayType`. Since they all return `Collection` values, we want to avoid nesting them.

We still support `Select 1 where 1 = ANY(Select collect_set(id) from t) which now wraps the `SetType` into a `TableType`. It's not totally obvious, we should add a comment.

File: core/src/main/java/io/crate/types/TableType.java (hunk 10 above)
Comment (mxm) [inline core/src/main/java/io/crate/types/TableType.java]: That's correct. However, we can extend the TableType to handle multiple columns in the future. We would then store an array of DataTypes in the TableType.

**Comment by mxm** [comment]
@mikethebeer The Exception message is still the same: `SQLActionException[UnsupportedFeatureException: Subquery returned more than 1 row]`

**Comment by mxm** [comment]
Opening a new PR which addresses the shortcomings here.
